cpp/src/plasma/client.cc

```diff
@@ -980,33 +984,59 @@ Status PlasmaClient::Impl::Subscribe(int* fd) {
   return Status::OK();
 }
 
-Status PlasmaClient::Impl::DecodeNotification(const uint8_t* buffer, ObjectID* object_id,
-                                              int64_t* data_size,
-                                              int64_t* metadata_size) {
+Status PlasmaClient::Impl::GetNotification(int fd, ObjectID* object_id,
+                                           int64_t* data_size, int64_t* metadata_size) {
   std::lock_guard<std::recursive_mutex> guard(client_mutex_);
 
-  auto object_info = flatbuffers::GetRoot<fb::ObjectInfo>(buffer);
-  ARROW_CHECK(object_info->object_id()->size() == sizeof(ObjectID));
-  memcpy(object_id, object_info->object_id()->data(), sizeof(ObjectID));
-  if (object_info->is_deletion()) {
-    *data_size = -1;
-    *metadata_size = -1;
-  } else {
-    *data_size = object_info->data_size();
-    *metadata_size = object_info->metadata_size();
+  if (pending_notification_.empty()) {
+    auto message = ReadMessageAsync(fd);
+    if (message == NULL) {
+      return Status::IOError("Failed to read object notification from Plasma socket");
+    }
+
+    std::vector<ObjectID> object_ids;
+    std::vector<int64_t> data_sizes;
+    std::vector<int64_t> metadata_sizes;
+    RETURN_NOT_OK(
+        DecodeNotifications(message.get(), &object_ids, &data_sizes, &metadata_sizes));
+    for (size_t i = 0; i < object_ids.size(); ++i) {
+      pending_notification_.push_back(
```

Comment:
You can actually call `pending_notification_.emplace_back(object_ids[i], data_sizes[i], metadata_sizes[i])` here AFAIK (it will get rid of the temporary copy.